---
title-block-banner-color: "#697ddd"
title: "DNA methylation desing to discovery 2024"
output: html_notebook
author: Dr. Gilles Gasporoni, Dr. Thomas Hentrich, Nihit Aggarwal
format:
  html:
    toc: true
    toc-location: left
---

## Day 1

### Setting up the environment

#### Installing packages

In this section, we will install RnBeads package from Bioconductor, essential for analyzing Infinium methylation arrays.

```{r,eval=FALSE}
# installing RnBeads 
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("FDb.InfiniumMethylation.hg19")

source("http://rnbeads.org/data/install.R")
```

#### Loading packages

we specify a list of essential R packages, including ggplot2, reshape2, RnBeads, and data.table, which are fundamental for data visualization and analysis in epigenetics studies. The lapply function is then used to load each package in the list, ensuring they are available for use in subsequent analyses.

```{r}
packages_to_load <- c("ggplot2","reshape2","RnBeads","data.table")
lapply(packages_to_load, require, character.only=TRUE)

```

### Primary analysis

#### Data and samplesheet preparation

In this segment, we are setting up the file paths and directories necessary for handling DNA methylation data in our epigenetics workshop. The code defines the locations for the raw data (`idat.dir`), the output directory (`report.dir`), and the sample sheet (`sample.annotation`), preparing us to import and analyze the data effectively in subsequent steps.

```{r}
# preparing sample sheet and data import for selected samples 

# set the working directory
setwd("D:/Work/Epigenetics/Workshops/2024_deNBI-DNA_methylation_design_to_discovery/")

# path to data.files
idat.dir <- file.path("D:/Work/Epigenetics/Workshops/2024_deNBI-DNA_methylation_design_to_discovery/idat/")

# outputPath
report.dir <- "D:/Work/Epigenetics/Workshops/2024_deNBI-DNA_methylation_design_to_discovery/analysis/"

# samplesheet file path
sample.annotation <- file.path("D:/Work/Epigenetics/Workshops/2024_deNBI-DNA_methylation_design_to_discovery/20241202_SampleSheet_SelectedSamples.txt")

data.type<-"idat.dir"

```

We first use the `help("rnb.options")` function to display detailed information about configurable parameters available in the RnBeads package. Following this, we configure several RnBeads options, setting the genome assembly to "hg19," specifying the identifiers column as "SampleID," and defining the import table separator as a tab. Finally, we export these configurations to an XML format with improved readability using `rnb.options2xml(pretty=TRUE)`. This setup ensures our RnBeads analyses are tailored to our specific dataset and preferences.

```{r}
# get info on all available RnB parameters.
help("rnb.options") 

# setting rnbeads options
rnb.options(assembly = "hg19")
rnb.options(identifiers.column = "SampleID")
rnb.options(import.table.separator = "\t")
rnb.options2xml(pretty=TRUE)
```

#### RnBeads set

The code initializes the RnBeads analysis by importing DNA methylation data and associated sample annotations from specified directories, subsequently storing the resulting RnBeads dataset object for further analysis.

```{r}
# import rnbead set
rnbSet = rnb.run.import(data.source = c(idat.dir, sample.annotation), 
                        data.type = data.type,
                        dir.reports = report.dir)
rnb = rnbSet$rnb.set
```

#### Features of RnBeads object

In this section, we explore key features of the RnBeads dataset to understand its structure and contents. The code first displays the top entries of the sample sheet using `pheno(rnb)`, then extracts raw methylation values with `meth(rnb)`, and finally visualizes the methylation distribution for a specific sample using a histogram. \##### Inspecting Dataset Features

```{r}
# Let's inspect some features of our dataset

# View Sample Sheet
head(pheno(rnb))

# Retrieve Raw Methylation Values
mm <- meth(rnb)

# Plot Raw Methylation Values Histogram
hist(mm[,5], col="steelblue", breaks=50)

# Check if Regions are Summarized and How
summarized.regions(rnb)

```

##### Working with Annotations

```{r}
# Inspect Available Annotation for Probes
anno = rnb.annotation2data.frame(rnb.get.annotation("probes450"))
head(anno)

# View Annotation Available for Promoters
annot.promoters <- annotation(rnb, type="promoters")
head(annot.promoters)
```

##### Detailed Methylation Analysis

```{r}
# get methylation for promoters (on a subset of samples and promoters just for illustration)
meth(rnb, type="promoters", row.names=TRUE, i=1:5, j=1:3)

# inspect coverage (number of beads per probe)
nbead <- covg(rnb, row.names=TRUE)
nbead[1:5,1:3]

# detection p-values
pvals <- dpval(rnb, row.names=TRUE)

```

##### Diagnostic Plots

```{r}
# check control probes
rnb.plot.control.boxplot(rnb)

# some specific control
rnb.plot.control.boxplot(rnb, "BISULFITE CONVERSION I")

# Negative control boxplots are generated with the following command:
rnb.plot.negative.boxplot(rnb)

# barplot of a selected control probe
control.meta.data <- rnb.get.annotation("controls450")
ctrl.probe<-paste0(unique(control.meta.data[["Target"]])[2], ".1")
rnb.plot.control.barplot(rnb, ctrl.probe)

# access to full qc results
qc_data = qc(rnb)
```

##### Genotyping and Additional Analysis

```{r}
#use some genotyping probes to check for genomic sample similarity
snp.probes = anno[grep("rs", rownames(anno)), ]

rnb.plot.snp.heatmap(rnb)
#rnb.plot.snp.barplot(dataset = rnb, probeID = rownames(snp.probes) [2], writeToFile = TRUE)

#set some options as desired before running the complete quality check module
rnb.options(qc.snp.boxplot=TRUE)
rnb.options(import.sex.prediction = TRUE)
rnb.options(qc.cnv = TRUE) # calculate copy number variations
```

#### Quality check

```{r}
#see all available options
help("rnb.options") #get info on all available RnB parameters.

#run complete quality report with specific settings

#QC
rnb.options(qc.boxplots = TRUE)
rnb.options(qc.barplots = TRUE)
rnb.options(qc.negative.boxplot = TRUE)
rnb.options(qc.snp.distances = TRUE)
rnb.options(qc.snp.boxplot = TRUE)
rnb.options(qc.snp.barplot = TRUE)
rnb.options(qc.sample.batch.size = 50)
rnb.options(qc.coverage.plots = FALSE)
rnb.options(qc.coverage.threshold.plot = 1:10)
rnb.options(qc.coverage.histograms = FALSE)
rnb.options(qc.coverage.violins = FALSE)

rnb.run.qc(rnb.set = rnb, dir.reports = report.dir) #this will takes some minutes
```

### Preprocessing
The following R code outlines the preprocessing steps for DNA methylation data using the RnBeads package. It begins by backing up the original unprocessed dataset. Various filtering techniques are then applied to remove undesirable data points based on criteria such as CpG context, SNP presence, and data quality (e.g., missing or low-variability values). Each filtering step concludes with a check on the remaining data, ensuring the refined dataset is ready for further analysis.

```{r}
# preprocessing typically includes filtering of undesired datapoints (mostly sites) and data normalization
rnb.unprocessed = rnb #first we store the unprocessed rnbeads object as a back-up
save.rnb.set(rnb.unprocessed, "/projects/researchers/jil/projects/deNBI/20241129_WS24/analysis/rnb/rnbSet_Unprocessed", archive = TRUE)

nrow(meth(rnb.unprocessed)) # the number of sites in the unfiltered object

#some examples of how sites to be removed from dataset
# Remove probes outside of CpG context
rnb.set.filtered <- rnb.execute.context.removal(rnb.set = rnb.unprocessed, contexts = NULL )$dataset
nrow(meth(rnb.set.filtered)) # the number of CpG sites in the unfiltered object

# SNP filtering allowing no SNPs in the probe sequence (range 3 bp)
rnb.set.filtered <- rnb.execute.snp.removal(rnb.set = rnb.set.filtered, snp = "3")$dataset

# Removal of CpG sites in the unfiltered object
# that contain a SNP in the range of 3bp
nrow(meth(rnb.set.filtered))

# Remove CpGs on sex chromosomes
rnb.set.filtered <- rnb.execute.sex.removal(rnb.set = rnb.set.filtered)$dataset
nrow(meth(rnb.set.filtered))

# Remove probes and samples based on a greedycut approach
greedycut.results <- rnb.execute.greedycut(rnb.set = rnb.set.filtered, pval.threshold = 0.05)#$dataset
to_remove = rownames(meth(object = rnb.set.filtered, row.names = TRUE)) [greedycut.results[["sites"]]]
remove.sites(object = rnb.set.filtered, probelist = to_remove)
nrow(meth(rnb.set.filtered))

# Remove probes containing NA for beta values
rnb.set.filtered <- rnb.execute.na.removal(rnb.set.filtered)$dataset
nrow(meth(rnb.set.filtered))

# Remove probes for which the beta values have low standard deviation
rnb.set.filtered <- rnb.execute.variability.removal(rnb.set.filtered, 0.005)$dataset
nrow(meth(rnb.set.filtered))

# we remove sites with any NA or negative values
mm = meth(rnb.set.filtered, row.names = TRUE)
mneg = apply(mm, 1, function(x) any(x <= 0))
mneg = is.na(mneg)
head(mm[mneg, 1:3],10)
rnb.set.filtered = remove.sites(object = rnb.set.filtered, probelist = mneg)
nrow(meth(rnb.set.filtered))
save.rnb.set(rnb.set.filtered, "/projects/researchers/jil/projects/deNBI/20241129_WS24/analysis/rnb/rnbSet_Filtered", archive = TRUE)
```


### Normalization
We now configure a series of RnBeads package options for filtering and normalization of DNA methylation data. It specifies parameters such as SNP filtering, cross-reactivity exclusion, and the handling of missing values and coverage thresholds. Additionally, it sets up normalization methods, including background correction and method adjustment. The final command executes the predefined preprocessing steps on the filtered dataset and stores the processed data in the `rnb` object for further analysis.
```{r}
#preset some options and then run filtering/normalization with one command
rnb.options(filtering.whitelist = NULL)
rnb.options(filtering.blacklist = NULL)
rnb.options(filtering.snp = "3")
rnb.options(filtering.cross.reactive = FALSE)
rnb.options(filtering.greedycut = TRUE)
rnb.options(filtering.greedycut.pvalue.threshold = 0.05)
rnb.options(filtering.greedycut.rc.ties = "row")
rnb.options(filtering.sex.chromosomes.removal = TRUE)
rnb.options(filtering.missing.value.quantile = 0.8)
rnb.options(filtering.coverage.threshold = 3)
rnb.options(filtering.low.coverage.masking = FALSE)
rnb.options(filtering.high.coverage.outliers = FALSE)
rnb.options(filtering.deviation.threshold = 0)

rnb.options(normalization = NULL)
rnb.options(normalization.method = "bmiq")       #normalization method
rnb.options(normalization.background.method = "methylumi.noob")#background removal yes/no
rnb.options(normalization.plot.shifts = TRUE)

#rnb.set.norm <- rnb.execute.normalization(rnb.set.unfiltered, method="wm.dazen", bgcorr.method="methylumi.noob")
preprocessed = rnb.run.preprocessing(rnb.set = rnb.set.filtered, dir.reports = report.dir) 
rnb = preprocessed$rnb.set
```


#inference module
```{r}
#Tissue deconvolution - reference based celltype estimation using houseman algorithm
ct <- rnb.execute.ct.estimation(rnb, cell.type.column="CellType", test.max.markers=10000, top.markers=500)
rnb.plot.ct.heatmap(ct.obj = ct)
ct$contributions

#immune cell content
immune.content <- rnb.execute.lump(rnb)

#calculate epigenetic age
rnb.execute.age.prediction(object = rnb)

#preset some settings
rnb.options(inference.age.prediction = TRUE)
rnb.options(inference.age.column = "Age")
rnb.options(inference.age.prediction.training = FALSE)
rnb.options(inference.age.prediction.cv = FALSE)
rnb.options(inference.immune.cells = TRUE)
rnb.options(inference.genome.methylation = "Genome-wide methylation")
rnb.options(inference.targets.sva = character())
rnb.options(inference.reference.methylome.column = "CellType")
rnb.options(inference.max.cell.type.markers = 10000)
rnb.options(inference.top.cell.type.markers = 500)
rnb.options(inference.sva.num.method = "leek")

rnb_inference = rnb.run.inference(rnb.set = rnb, dir.reports = report.dir)
rnb = rnb_inference$rnb.set

#let's do the sva analysis as an addendum
sva.obj <- rnb.execute.sva(rnb, cmp.cols = "Group", numSVmethod="be")
s
rnb <- set.covariates.sva(rnb, sva.obj)

#####################################################################

#exploratory module
rnb.run.exploratory(rnb.set = rnb, dir.reports = report.dir)

Sys.setenv("R_ZIPCMD"="/usr/bin/zip")
save.rnb.set(rnb, "/projects/researchers/jil/projects/deNBI/20241129_WS24/analysis/rnb/rnbSet_exploratoryAnalysis", archive = TRUE)

# load.rnb.set

#####################################################################


```

## Day 2

### 

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
